### PR TITLE
SIMSBIOHUB-587 API: Marking Strategy

### PR DESCRIPTION
**Pull request by MacDeluca**
## Links to Jira Tickets

- {Include a link to all applicable Jira tickets}

## Description of Changes

- {List all relevant code changes. Include any changes to the business workflow that might not be obvious to the reviewers of this PR.}

## Testing Notes

- File in discord

### CODE REVIEW

File: api/src/services/critterbase-service.ts [hunk 1]
```diff
@@ -174,7 +204,7 @@ export interface IBulkCreate {
   collections?: ICollection[];
   mortalities?: IMortality[];
   locations?: ILocation[];
-  markings?: IMarking[];
+  markings?: IMarking[] | IBulkCreateMarking[];
```
Comment (NickPhura) [inline api/src/services/critterbase-service.ts:207]: Just checking: This `IBulkCreate` interface is used by both the `updateCritter` and `bulkCreate` functions in `critterbase-service.ts`. Are they both okay with receiving either `IMarking[]` or `IBulkCreateMarking[]`?

File: api/src/services/import-services/marking/import-markings-strategy.test.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import sinon from 'sinon';
+import sinonChai from 'sinon-chai';
+import { getMockDBConnection } from '../../../__mocks__/db';
+import { IBulkCreateResponse, ICritterDetailed } from '../../critterbase-service';
+import { ImportMarkingsStrategy } from './import-markings-strategy';
+import { CsvMarking } from './import-markings-strategy.interface';
+
+chai.use(sinonChai);
+
+describe.only('ImportMarkingsStrategy', () => {
```
Comment (NickPhura) [inline api/src/services/import-services/marking/import-markings-strategy.test.ts]: .only

File: api/src/services/import-services/utils/datetime.ts [hunk 3]
```diff
@@ -25,3 +32,45 @@ export const formatTimeString = (time?: string): string | undefined => {
...
+ * @param {string} _dateA - Date string
+ * @param {string} _dateB - Date string
+ * @returns {string | undefined}
+ */
+export const areDatesEqual = (_dateA: string, _dateB: string): boolean => {
+  const format = 'dd-mm-yyyy';
+  const dateA = dayjs(_dateA, format);
+  const dateB = dayjs(_dateB, format);
+
+  return dateA.isValid() && dateB.isValid() && dateA.format(format) === dateB.format(format);
```
Comment (NickPhura) [inline api/src/services/import-services/utils/datetime.ts]: dayjs does have an `isSame` function, if that works here. 

https://day.js.org/docs/en/query/is-same

Comment (NickPhura) [reply api/src/services/import-services/utils/datetime.ts:51]: Though, the equals might also just do the same thing, if its just comparing them as strings

File: api/src/services/import-services/marking/import-markings-strategy.test.ts (hunk 2 above)
Comment (MacDeluca) [reply api/src/services/import-services/marking/import-markings-strategy.test.ts]: oops fixed

File: api/src/services/critterbase-service.ts (hunk 1 above)
Comment (MacDeluca) [reply api/src/services/critterbase-service.ts:207]: I wrote a TODO at the top of this file kinda referencing this. It looks like a lot of the types in this file are being used as request bodies AND response bodies which is pretty messy. 

I think we should go through and audit this service / clean up some of the strange types.

I think also simplifying the strange constant mapping for endpoints and use a better query param parser.

Comment (MacDeluca) [reply api/src/services/critterbase-service.ts:207]: The ```updateCritter``` method is actually just hitting the bulkUpdate endpoint

File: api/src/services/import-services/utils/datetime.ts (hunk 3 above)
Comment (MacDeluca) [reply api/src/services/import-services/utils/datetime.ts]: Updated.
Yea I think they work the same but the isSame version is cleaner.

File: api/src/services/import-services/utils/datetime.test.ts [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import { expect } from 'chai';
-import { formatTimeString } from './datetime';
+import { areDatesEqual, formatTimeString } from './datetime';
 
 describe('formatTimeString', () => {
```
Comment (NickPhura) [inline api/src/services/import-services/utils/datetime.test.ts:4]: Could add a unit test for when the input is `null`

File: api/src/services/import-services/marking/import-markings-strategy.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+   * @returns {*}
+   */
+  async validateRows(rows: Row[]) {
+    // Generate type-safe cell getter from column validator
+    const getCellValue = generateCellGetterFromColumnValidator(this.columnValidator);
+
+    // Get validation reference data
+    const critterAliasMap = await this.surveyCritterService.getSurveyCritterIdAliasMap(this.surveyId);
+    const colours = await this.surveyCritterService.critterbaseService.getColours();
+    const markingTypes = await this.surveyCritterService.critterbaseService.getMarkingTypes();
```
Comment (NickPhura) [inline api/src/services/import-services/marking/import-markings-strategy.ts]: super nit: could Promise.all these 3 calls

File: api/src/paths/project/{projectId}/survey/{surveyId}/critters/markings/import.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+      const importCsvMarkingsStrategy = new ImportMarkingsStrategy(connection, surveyId);
+
+      // Pass CSV file and importer as dependencies
+      const markingsCreated = await importCSV(parseMulterFile(rawFile), importCsvMarkingsStrategy);
+
+      await connection.commit();
+
+      return res.status(201).json({ markingsCreated });
+    } catch (error) {
+      defaultLog.error({ label: 'importCritterCsv', message: 'error', error });
```
Comment (NickPhura) [inline api/src/paths/project/{projectId}/survey/{surveyId}/critters/markings/import.ts]: mega nit: error label doesn't match the function name

File: api/src/services/survey-critter-service.ts [hunk 7]
```diff
@@ -249,34 +249,36 @@ export class SurveyCritterService extends DBService {
...
+        const alias = critter.animal_id.toLowerCase();
+        // Do not allow existing duplicate aliases to map over eachother
+        if (critterAliasMap.get(alias)) {
+          defaultLog.debug({
+            label: 'critter alias map',
+            message: 'duplicate critter alias for survey',
+            details: { surveyId: surveyId, critter: critter }
+          });
+
+          critterAliasMap.set(alias, undefined);
```
Comment (NickPhura) [inline api/src/services/survey-critter-service.ts:281]: Is this how you trigger the zod error? Maybe worth adding a comment just in case

Comment (NickPhura) [reply api/src/services/survey-critter-service.ts:281]: I guess the log above kind of describes the situation, though a separate comment here might still be safest

File: api/src/services/import-services/utils/datetime.test.ts (hunk 4 above)
Comment (MacDeluca) [reply api/src/services/import-services/utils/datetime.test.ts:4]: roger

File: api/src/services/import-services/marking/import-markings-strategy.ts (hunk 5 above)
Comment (MacDeluca) [reply api/src/services/import-services/marking/import-markings-strategy.ts]: I like it 🦟

File: api/src/paths/project/{projectId}/survey/{surveyId}/critters/markings/import.ts (hunk 6 above)
Comment (MacDeluca) [reply api/src/paths/project/{projectId}/survey/{surveyId}/critters/markings/import.ts]: damnit

File: api/src/services/survey-critter-service.ts (hunk 7 above)
Comment (MacDeluca) [reply api/src/services/survey-critter-service.ts:281]: updated

**Review by NickPhura — APPROVED** [review]
🐤
